precompile/modules/initia_stdlib/sources/vip/vesting.move

```diff
@@ -323,451 +287,625 @@ module initia_std::vip_vesting {
+        vip_reward::register_reward_store<OperatorVesting>(chain, bridge_id)
+    }
+
+    public(friend) fun supply_reward_on_user(
+        bridge_id: u64,
+        stage: u64,
+        reward: FungibleAsset,
+    ) {
+        let reward_store_addr = get_user_reward_store_address(bridge_id);
+        vip_reward::add_reward_per_stage(
+            reward_store_addr,
+            stage,
+            fungible_asset::amount(&reward)
+        );
+        primary_fungible_store::deposit(reward_store_addr, reward);
+    }
+
+    public(friend) fun supply_reward_on_operator(
+        bridge_id: u64,
+        stage: u64,
+        reward: FungibleAsset,
+    ) {
+        let reward_store_addr = get_operator_reward_store_address(bridge_id);
+        vip_reward::add_reward_per_stage(
+            reward_store_addr,
+            stage,
+            fungible_asset::amount(&reward)
+        );
+        primary_fungible_store::deposit(reward_store_addr, reward);
+    }
+
+    // calculate user vesting until previous stage to
+    fun batch_claim_previous_user_vestings(
+        account_addr: address,
+        bridge_id: u64,
+        user_vestings: &mut vector<UserVesting>,
+        vesting_store: &mut VestingStore<UserVesting>,
+        penalty_keys: &mut vector<u64>,
+        penalty_map: &mut SimpleMap<u64, u64>,
+        claim_info: &UserVestingClaimInfo
+    ): (u64, u64) {
+        let vested_reward = 0u64;
+        let penalty_reward = 0u64;
+        let finalized_vestings_idx: vector<u64> = vector[]; // vector index
+        let idx = 0;
+        let len = vector::length(user_vestings);
+        while (idx < len) {
+            let value = vector::borrow_mut(user_vestings, idx);
+            let vest_amount = if (claim_info.l2_score >= value.minimum_score) {value.vest_max_amount} else {
+                (
+                    (value.vest_max_amount as u128) * (claim_info.l2_score as u128) / (
+                        value.minimum_score as u128
+                    ) as u64
+                )
+            };
+            vested_reward = vested_reward + vest_amount;
+            penalty_reward = penalty_reward + value.vest_max_amount - vest_amount;
+            value.remaining_reward = value.remaining_reward - value.vest_max_amount;
+
+            if (penalty_reward > 0) {
+                if (!simple_map::contains_key(penalty_map, &value.start_stage)) {
+                    simple_map::add(penalty_map, value.start_stage, 0);
+                    vector::push_back(penalty_keys, value.start_stage);
+                };
+
+                let penalty_amount = simple_map::borrow_mut(penalty_map, &value.start_stage);
+                *penalty_amount = *penalty_amount + penalty_reward;
+            };
 
-            // move vesting if end stage is over or the left reward is empty
-            if (stage > value.end_stage || value.remaining_reward == 0) {
+            if (claim_info.start_stage >= value.end_stage) {
```

Comment:
I think this check should be done right after we borrow vesting value and do emit event and `continue` if the stage is out of range and change this line to just checking `claim_info.start_stage == value.end_stage`.